### PR TITLE
[lit-html] Handle <template> elements in templates

### LINKED ISSUE #1737: [lit-html] Template literals with parts inside `<template>`s render incorrectly.
# Description

When rendering a template string where some of the parts are inside of a `<template>`, values are rendered into the wrong parts. ([`TemplateInstance#clone()`](https://github.com/Polymer/lit-html/blob/2be1fcf0a82a2bf368295d015613cd87c5ad96f6/packages/lit-html/src/lit-html.ts#L750) doesn't walk `<template>` content to find parts.)

# lit-html version

2.0.0-pre.7

# Steps to Reproduce

1. Write this code

```ts
import {html, render} from 'https://unpkg.com/lit-html@2.0.0-pre.7?module';
render(html`${'A'}<template>${'B'}</template>${'C'}`, document.body);
```

2. See this output...

```
AB
```

## Live Reproduction Link
https://jsbin.com/xojudijari/edit?html,output

## Expected Results
The page shows `AC`. (`'B'` should be rendered to a part inside a `<template>`.)

## Actual Results
The page shows `AB`.

### PR DESCRIPTION
**Pull request by justinfagnani**
Fixes #1737

### CODE REVIEW

File: packages/tests/web-test-runner.config.js [hunk 1]
```diff
@@ -30,8 +30,8 @@ const browserPresets = {
   // Default set of Playwright browsers to test when running locally.
   local: [
     'chromium', // keep browsers on separate lines
-    'firefox', // to make it easier to comment out
-    'webkit', // individual browsers
+    // 'firefox', // to make it easier to comment out
+    // 'webkit', // individual browsers
```
Comment (kevinpschaaf) [inline packages/tests/web-test-runner.config.js:34]: Remove?

File: packages/tests/web-test-runner.config.js [hunk 2]
```diff
@@ -198,7 +198,7 @@ export default {
     }),
   ],
   // Only actually log errors and warnings. This helps make test output less spammy.
-  filterBrowserLogs: (type) => type === 'warn' || type === 'error',
+  // filterBrowserLogs: (type) => type === 'warn' || type === 'error',
```
Comment (kevinpschaaf) [inline packages/tests/web-test-runner.config.js:201]: Remove?

File: packages/lit-html/package.json [hunk 3]
```diff
@@ -58,7 +58,7 @@
     "dev": "scripts/dev.sh",
     "prepublishOnly": "npm run check-version",
     "test": "npm run test:dev && npm run test:prod",
-    "test:dev": "cd ../tests && npx wtr '../lit-html/development/**/*_test.(js|html)'",
+    "test:dev": "cd ../tests && npx wtr '../lit-html/development/**/lit-html_test.js'",
```
Comment (kevinpschaaf) [inline packages/lit-html/package.json:61]: Remove?

And note you can just do `npm run test:dev -- -- '../lit-html/development/**/lit-html_test.js'` on the command line to focus on one test if you don't want to use the CLI GUI for that.

File: packages/lit-html/src/test/lit-html_test.ts [hunk 4]
```diff
@@ -391,24 +391,12 @@ suite('lit-html', () => {
...
-      );
-
-      // Currently fails:
-      // render(html`<${'A'}></ ${'A'}>${'B'}`, container);
-      // assert.equal(stripExpressionMarkers(container.innerHTML), '<><!-- -->B</>');
+      assert.throws(() => render(html`<${'A'}></ ${'A'}>`, container));
     });
 
-    test('binding after end tag name', () => {
+    test.skip('binding after end tag name', () => {
```
Comment (kevinpschaaf) [inline packages/lit-html/src/test/lit-html_test.ts:399]: Remove skip?

**Review by kevinpschaaf — COMMENTED** [review]
This looks like it regresses benchmarks on top of code size.

I think I'd prefer marking this as a breaking change, and see if we get strong feedback on people having relied on this for "interesting" use cases enough to justify. The only one we've seen is using it to put comments in CSS in polymer templates via binding expressions (!!!), and for that, they could use `static` expressions.

**Comment by graynorton** [comment]
Just noting that someone called out a specific use case that's not working: https://github.com/lit/lit/issues/1737#issuecomment-829224907

**Comment by madfriend** [comment]
Sorry to be straightforward, but is there any chance on this PR getting merged? I stumbled upon the bug of nested `<template>` and it blocks the migration of our codebase from lit-html@1.2.1 to lit@2.

**Comment by justinfagnani** [comment]
@madfriend we aren't sure. This has a negative perf impact on all templates, whether or not they have bindings in a `<template>` tag. We are considering suggesting that you use the [static expression support](https://lit.dev/docs/templates/expressions/#static-expressions) as a replacement for this functionality, since it scopes the perf impact to one templates.

**Comment by madfriend** [comment]
@justinfagnani Are there any insights on why does it affect performance of all templates, not just the ones with `<template>` ? If I understood the suggested change correctly, we only increase the frontier of examined parts when we come across the template tag, otherwise the logic is pretty much the same. 

As for the static expressions, it just seems odd to use them as a workaround because internally lit-html stopped iterating over `<template>` children. It would mostly certainly be not easy to explain this to newcomers.

**Comment by justinfagnani** [comment]
<template> "children" aren't really children - they're moved to a separate property that we have to check for, so the additional code path and conditionals to check whether the element is a template effect everyone, even if we don't recurse into the template content.

One reason why static makes sense for <template> is that changing the bound values in expressions in <templates> is of limited use - existing DOM created from those templates will not be updated, only the template itself. So if those values usually shouldn't and don't change, static templates optimize for that.

**Comment by madfriend** [comment]
I see your point, but I am concerned that this is a breaking change for those who migrate from lit-html@1.2.1 to lit@2. Shouldn't this change be at least described in the docs (somewhere along the point about expressions in comments)?

**Comment by justinfagnani** [comment]
Yes, I thought we had it listed here: https://lit.dev/docs/releases/upgrade/#lit-html but it look like we missed it.

**Comment by justinfagnani** [comment]
I'm going to close this as too costly in size and perf for the majority of Lit templates that won't be binding inside `<template>`. We're going to recommend using static templates and will write docs to show that.

**Pull request closed without merging**